server/src/main/java/org/opensearch/action/ActionModule.java

```diff
@@ -954,13 +964,107 @@ protected void configure() {
+        return dynamicActionRegistry;
+    }
+
     public RestController getRestController() {
         return restController;
     }
+
+    /**
+     * The DynamicActionRegistry maintains a registry mapping {@link ExtensionAction} instances to {@link TransportAction} instances.
+     * <p>
+     * This class is modeled after {@link NamedRegistry} but provides both register and unregister capabilities.
+     *
+     * @opensearch.internal
+     */
+    public static class DynamicActionRegistry {
+        // This is the immutable actions map created during node bootstrap, which
+        // will continue to link ActionType and TransportAction pairs from core and plugin
+        // action handler registration.
+        private Map<ActionType, TransportAction> actions = Collections.emptyMap();
+        // A dynamic registry to add or remove ExtensionAction / ExtensionTransportAction
+        // pairs. These classes extend ActionType and TransportAction, respectively, with
+        // additional fields identifying the remote extension they are associated with.
+        private final Map<ExtensionAction, ExtensionTransportAction> registry = new ConcurrentHashMap<>();
+
+        private ActionFilters actionFilters;
+        private ExtensionsManager extensionsManager;
+        private TransportService transportService;
+
+        /**
+         * Initialize the immutable actions in the registry and provide parameters needed for dynamic actions.
+         *
+         * @param actions The injected map of {@link ActionType} to {@link TransportAction}
+         * @param actionFilters The action filters
+         * @param transportService The node's {@link TransportService}
+         * @param extensionsManager The instance of the {@link ExtensionsManager}
+         */
+        public void initialize(
+            Map<ActionType, TransportAction> actions,
+            ActionFilters actionFilters,
+            TransportService transportService,
+            ExtensionsManager extensionsManager
+        ) {
+            this.actions = actions;
+            this.actionFilters = actionFilters;
+            this.transportService = transportService;
+            this.extensionsManager = extensionsManager;
+        }
+
+        /**
+         * Add an {@link ExtensionAction} to the registry.
+         *
+         * @param extensionAction The action to add
+         */
+        public void registerExtensionAction(ExtensionAction extensionAction) {
+            requireNonNull(extensionAction, "extension action is required");
+            String name = extensionAction.name();
+            String uniqueId = extensionAction.uniqueId();
+            if (registry.containsKey(extensionAction)) {
+                throw new IllegalArgumentException("extension [" + uniqueId + "] action [" + name + "] already registered");
+            }
+            registry.put(extensionAction, new ExtensionTransportAction(name, actionFilters, transportService, extensionsManager));
+        }
```

Comment:
hm... I have couple of concerns:
- ActionModule/DynamicActionRegistry shouldn't understand details of extension business logic. 
Do you think putting in an interface of registerActionFilter and let ExtensionManager implement its own filter which would cleanly implement its business logic. 

- This really doesnt support registering actions dynamically for rest of the system. 
Developers would love a way to register these actions dynamically. I can think of plugins leveraging this for hotswap, register actions based on settings toggled etc.

Reply:
This is just a naming/typing thing that's easily fixed.  `registerDynamicAction(ActionType action)` would do the same thing, and we could generalize the uniqueID to just some other id field.